pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,664 @@
+		return nil
+	}
+
+	quotaName := originQuota.Name
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+	localQuotaInfo, exist := qt.quotaInfoMap[quotaName]
+	if !exist {
+		err = fmt.Errorf("quota not exist:%v", quotaName)
+		klog.Errorf("UpdateQuota failed:%v", err)
+		return err
+	}
+
+	err = qt.basicItemCheck(originQuota)
+	if err != nil {
+		klog.Errorf("UpdateQuota basicItemCheck failed %v", err)
+		return err
+	}
+
+	newQuotaInfo := core.NewQuotaInfoFromQuota(originQuota)
+	// quota 信息检查
+	err = qt.validateQuota(newQuotaInfo)
+	if err != nil {
+		klog.Errorf("UpdateQuota validateQuota failed:%v", err)
+		return err
+	}
+
+	// quota 层次关系发生改变
+	err = qt.checkQuotaHierarchyHasChange(localQuotaInfo, newQuotaInfo)
+	if err != nil {
+		klog.Errorf("UpdateQuota checkQuotaHierarchyHasChange failed:%v", err)
+		return err
+	}
+
+	qt.quotaInfoMap[quotaName] = newQuotaInfo
+	return nil
+}
+
+func (qt *quotaTopology) DeleteQuota(quotaName string) (err error) {
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	quotaInfo, exist := qt.quotaInfoMap[quotaName]
+	if !exist {
+		return fmt.Errorf("not found quota:%v", quotaName)
+	}
+
+	// check has child quota
+	if childSet, exist := qt.quotaHierarchyInfo[quotaName]; exist {
+		if len(childSet) > 0 {
+			return fmt.Errorf("delete quota failed:%v, has child quota", quotaName)
+		}
+	} else {
+		return fmt.Errorf("BUG quotaMap and quotaTree information out of sync, losed :%v", quotaName)
+	}
+
+	// 从父quota中删除
+	if quotaInfo.ParentName != extension.RootQuotaName {
+		delete(qt.quotaHierarchyInfo[quotaInfo.ParentName], quotaName)
+	}
+	// 从拓扑结构中删除
+	delete(qt.quotaHierarchyInfo, quotaName)
+	delete(qt.quotaInfoMap, quotaName)
+	delete(qt.podCache, quotaName)
+	return nil
+}
+
+func (qt *quotaTopology) ValidateDeleteQuota(quota *v1alpha1.ElasticQuota) (err error) {
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	quotaName := quota.Name
+	info, exist := qt.quotaInfoMap[quotaName]
+	if !exist {
+		return fmt.Errorf("not found quota:%v", quotaName)
+	}
+
+	// check has child quota
+	if childSet, exist := qt.quotaHierarchyInfo[quotaName]; exist {
+		if len(childSet) > 0 {
+			return fmt.Errorf("delete quota failed:%v, has child quota", quotaName)
+		}
+	} else {
+		return fmt.Errorf("BUG quotaMap and quotaTree information out of sync, losed :%v", quotaName)
+	}
+
+	if !info.IsParent {
+		if len(qt.podCache[quotaName]) > 0 {
+			return fmt.Errorf("delete quota failed:%v, has linked pods", quotaName)
+		}
+	}
+	return nil
+}
+
+func (qt *quotaTopology) basicItemCheck(quota *v1alpha1.ElasticQuota) (err error) {
+	if invalid, err := extension.IsForbiddenModify(quota); invalid {
+		return err
+	}
+
+	if parentName, exist := quota.Labels[extension.LabelQuotaParent]; exist && parentName != extension.RootQuotaName {
+		parentInfo, find := qt.quotaInfoMap[parentName]
+		if !find {
+			err = fmt.Errorf("%v has parentName %v but not found parentInfo", quota.Name, parentName)
+			return err
+		}
+		if parentInfo.Name != parentName {
```

Comment:
if parentInfo.Name != parentName {} only happen with your code bug, kill this judge